### PR TITLE
Case insensitive username and email indexing and query planning for Postgres

### PR DESCRIPTION
**Pull request by cbaker6**
Initial code, still needs to be fixed

### CODE REVIEW

**Comment by dplewis** [comment]
@cbaker6 Can you check the travis build? You need to run the linter.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -58,20 +58,20 @@ Once you have babel running in watch mode, you can start making changes to parse
 
 If your pull request introduces a change that may affect the storage or retrieval of objects, you may want to make sure it plays nice with Postgres.
 
-* Run the tests against the postgres database with `PARSE_SERVER_TEST_DB=postgres npm test`. You'll need to have postgres running on your machine and setup [appropriately](https://github.com/parse-community/parse-server/blob/master/.travis.yml#L37) or use [`Docker`](#run-a-parse-postgres-with-docker).
+* Run the tests against the postgres database with `PARSE_SERVER_TEST_DB=postgres npm test`. You'll need to have postgres running on your machine and your configuratoin file setup [appropriately](https://github.com/postgres/postgres/blob/master/src/backend/utils/misc/postgresql.conf.sample#L63) along with [PostGIS](https://postgis.net) 2.2.0 higher installed, or use [`Docker`](#run-a-parse-postgres-with-docker).
```
Comment (dplewis) [inline CONTRIBUTING.md]: Since the build script has been moved from travis you should link to it.

https://github.com/parse-community/parse-server/blob/master/scripts/before_script_postgres.sh#L9

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -58,20 +58,20 @@ Once you have babel running in watch mode, you can start making changes to parse
...
 
 ```sh
-docker run -d --name parse-postgres -p 5432:5432 -e POSTGRES_USER=$USER --rm mdillon/postgis:11-alpine && sleep 5 && docker exec -it parse-postgres psql -U $USER -c 'create database parse_server_postgres_adapter_test_database;' && docker exec -it parse-postgres psql -U $USER -c 'CREATE EXTENSION postgis;' -d parse_server_postgres_adapter_test_database && docker exec -it parse-postgres psql -U $USER -c 'CREATE EXTENSION postgis_topology;' -d parse_server_postgres_adapter_test_database
+docker run -d --name parse-postgres -p 5432:5432 -e POSTGRES_PASSWORD=postgres --rm postgis/postgis:11-2.5-alpine && sleep 20 && docker exec -it parse-postgres psql -U $USER -c 'CREATE DATABASE parse_server_postgres_adapter_test_database;' && docker exec -it parse-postgres psql -U $USER -c 'CREATE EXTENSION postgis;' -d parse_server_postgres_adapter_test_database && docker exec -it parse-postgres psql -U $USER -c 'CREATE EXTENSION postgis_topology;' -d parse_server_postgres_adapter_test_database
```
Comment (dplewis) [inline CONTRIBUTING.md]: Have you tested this in docker?

File: package.json [hunk 3]
```diff
@@ -37,6 +37,7 @@
     "graphql-relay": "^0.6.0",
     "graphql-tools": "^4.0.7",
     "graphql-upload": "10.0.0",
+    "http2": "^3.3.7",
```
Comment (dplewis) [inline package.json]: remove this

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 4]
```diff
@@ -621,11 +621,11 @@ const buildWhereClause = ({
       const distance = fieldValue.$maxDistance;
       const distanceInKM = distance * 6371 * 1000;
       patterns.push(
-        `ST_distance_sphere($${index}:name::geometry, POINT($${index +
+        `ST_DistanceSphere($${index}:name::geometry, POINT($${index +
```
Comment (dplewis) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: This should be in a separate PR.

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 5]
```diff
@@ -2523,9 +2523,42 @@ export class PostgresStorageAdapter implements StorageAdapter {
...
-  ensureIndex(): Promise<void> {
-    return Promise.resolve();
+  ensureIndex(
+    className: string,
+    schema: SchemaType,
+    fieldNames: string[],
+    indexName: ?string,
+    caseInsensitive: boolean = false
+  ): Promise<void> {
+    const defaultIndexName = `parse_default_${fieldNames.sort().join('_')}`;
```
Comment (dplewis) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Can you add tests cases for this?

File: package.json (hunk 3 above)
Comment (cbaker6) [reply package.json]: this was added by accident when installing locally... removed

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js (hunk 4 above)
Comment (cbaker6) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: switched back

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 6]
```diff
@@ -2523,9 +2523,41 @@ export class PostgresStorageAdapter implements StorageAdapter {
...
+  ): Promise<any> {
+    const defaultIndexName = `parse_default_${fieldNames.sort().join('_')}`;
+    const indexNameOptions: Object = indexName ? { name: indexName } : { name: defaultIndexName };
+    const constraintPatterns =  caseInsensitive ? fieldNames.map(
+      (fieldName, index) => `lower($${index + 3}:name) varchar_pattern_ops`
+    ) : fieldNames.map(
+      (fieldName, index) => `$${index + 3}:name`
+    );
+    const qs = `CREATE INDEX $1:name ON $2:name (${constraintPatterns.join()})`;
+    return this._client.none(qs, [indexName, className, ...fieldNames])
```
Comment (cbaker6) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: I think the issue with the test failing is how this returns. It adds the index, but I did something wrong with the promise

File: package.json (hunk 3 above)
Comment (dplewis) [reply package.json]: Can you revert the package-lock.json

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 7]
```diff
@@ -2523,9 +2523,41 @@ export class PostgresStorageAdapter implements StorageAdapter {
...
+          error.code === PostgresUniqueIndexViolationError &&
+          error.message.includes(indexNameOptions)
+        ) {
+          // Cast the error into the proper parse error
+          throw new Parse.Error(
+            Parse.Error.DUPLICATE_VALUE,
+            'A duplicate value for a field with unique values was provided'
+          );
+        } else {
+          throw error;
```
[the remark this answers is not in the document]
Comment (cbaker6) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: fixed

**Comment by dplewis** [comment]
@cbaker6 Can you fix the travis build? Also are you able to run the tests locally for PG?

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js (hunk 5 above)
Comment (cbaker6) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Done, added a test case that uses the default name

File: CONTRIBUTING.md (hunk 1 above)
Comment (cbaker6) [reply CONTRIBUTING.md]: I added to the section which shows how to run docker

File: CONTRIBUTING.md (hunk 2 above)
Comment (cbaker6) [reply CONTRIBUTING.md]: Updated and fixed

File: package.json (hunk 3 above)
Comment (cbaker6) [reply package.json]: Done

**Comment by cbaker6** [comment]
@dplewis @acinader I didn't make the change, but it seems like this [line](https://github.com/parse-community/parse-server/blob/c9dcab6530d08beda948b84ecd8d77691e9962ac/src/Adapters/Storage/Postgres/PostgresStorageAdapter.js#L289):

``` 
else if (
      caseInsensitive &&
      (fieldName === 'username' || fieldName === 'email')
    ) {
      patterns.push(`LOWER($${index}:name) = LOWER($${index + 1})`);
      values.push(fieldName, fieldValue);
      index += 2;
    }
```

Is limiting case sensitivity for where clauses to just username and email fields. The way I wrote "ensureIndex", any field can have a caseInsensitive index. Is it possible this line can be changed to?

```
else if (
      caseInsensitive
    ) {
      patterns.push(`LOWER($${index}:name) = LOWER($${index + 1})`);
      values.push(fieldName, fieldValue);
      index += 2;
    }
```

**Comment by cbaker6** [comment]
> @dplewis @acinader I didn't make the change, but it seems like these [line](https://github.com/parse-community/parse-server/blob/c9dcab6530d08beda948b84ecd8d77691e9962ac/src/Adapters/Storage/Postgres/PostgresStorageAdapter.js#L289):
> 
> ```
> else if (
>       caseInsensitive &&
>       (fieldName === 'username' || fieldName === 'email')
>     ) {
>       patterns.push(`LOWER($${index}:name) = LOWER($${index + 1})`);
>       values.push(fieldName, fieldValue);
>       index += 2;
>     }
> ```
> 
> Is limiting case sensitivity for where clauses to just username and email fields. The way I wrote "ensureIndex", any field can have a caseInsensitive index. Is it possible this line can be changed to?
> 
> ```
> else if (
>       caseInsensitive
>     ) {
>       patterns.push(`LOWER($${index}:name) = LOWER($${index + 1})`);
>       values.push(fieldName, fieldValue);
>       index += 2;
>     }
> ```

I ended up changing this and added a test case using a case sensitive objectId.

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 8]
```diff
@@ -2369,10 +2389,38 @@ export class PostgresStorageAdapter implements StorageAdapter {
           sort !== undefined && sorting.length > 0 ? `ORDER BY ${sorting}` : '';
       }
     }
-
-    const qs = `SELECT ${columns.join()} FROM $1:name ${wherePattern} ${sortPattern} ${limitPattern} ${skipPattern} ${groupPattern}`;
+    const originalQuery = `SELECT ${columns.join()} FROM $1:name ${wherePattern} ${sortPattern} ${limitPattern} ${skipPattern} ${groupPattern}`;
+    const qs = explain ? this.createExplainableQuery(originalQuery) : originalQuery;
     debug(qs, values);
     return this._client
+      /*.any(qs, values)
```
Comment (cbaker6) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: There's a bug here... Was trying to switch the mapping method to many so it will return the "explain" results here, but had trouble with converting to map/promise

Comment (cbaker6) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Fixed

**Comment by cbaker6** [comment]
On a related note and for future reference, I think the changes here helped me find the root cause of #6374. My assumption is that when previously running tests with postgres-9, the issue didn't show because postgres-9 is pre-installed on the image and runs on port 5432. The change to postgres 11 required an add-on and two issues were created:

1. Pre-installed versions of postgres run on port 5432
2. Using `sudo service postgresql start ${ADD_ON_VERSION}`

Looking at the raw logs of travis, when postgres installs as an add-on, it stops all postgres services and starts up the add-on version using: `sudo systemctl start postgresql@${ADD_ON_VERSION}-main`. It doesn't use "sudo service..." and my guess is there are some environment vars that aren't updated to point to the add-on version. This caused a conflict because something (I don't know what it was) was looking for postgis on the default port of 5432, but instead found the add-on version which had an updated version of postgis. 

So if you really want to run on 5432, change the port in the add-on as before, `sudo sed -i 's/port = 5433/port = 5432/' /etc/postgresql/${POSTGRES_MAJOR_VERSION}/main/postgresql.conf` and then start the service using `sudo service postgresql start ${ADD_ON_VERSION}` instead of ~~sudo service postgresql start ${ADD_ON_VERSION}~~.

Or, with the fix in this branch that fixes the hard coded URIs in two of the spec files, simply use the already included `PARSE_SERVER_TEST_DATABASE_URI` to set the URI to the default add-on port of 5433, `PARSE_SERVER_TEST_DATABASE_URI=postgres://localhost:5433/parse_server_postgres_adapter_test_database`

**Comment by cbaker6** [comment]
I updated the contributing guide for running postgres tests. The rest of the updates for how to run in docker are in pull request #6506

**Comment by cbaker6** [comment]
> @cbaker6 Can you fix the travis build? Also are you able to run the tests locally for PG?

I missed this comment, but yes I can run the postgres adapter tests locally. I couldn't run the whole spec locally because of the bugs in #6531. I run postgres via docker locally along with some remote servers I have deployed. Both fail, which led me to fixes in 6531. Now I can run locally and remotely

**Comment by acinader** [comment]
@dplewis you ok with this?

also want to see the tests pass as master is currently hanging on the postgres test.

**Comment by acinader** [comment]
ok.  tests look good....

**Comment by dplewis** [comment]
@cbaker6 Since https://github.com/parse-community/parse-server/pull/6531 is merged can you resolve the conflicts?
